pandas/core/case_when.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+from itertools import zip_longest
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from pandas._libs import lib
+
+from pandas.core.dtypes.cast import (
+    construct_1d_arraylike_from_scalar,
+    find_common_type,
+    infer_dtype_from,
+)
+from pandas.core.dtypes.common import is_scalar
+from pandas.core.dtypes.generic import ABCNDFrame
+
+from pandas.core.construction import array as pd_array
+
+if TYPE_CHECKING:
+    from pandas._typing import Series
+
+
+def case_when(
+    *args: tuple[tuple],
+    default=lib.no_default,
+    level: int | None = None,
+) -> Series:
+    """
+    Replace values where the conditions are True.
+    It is a convenience wrapper around `pd.Series.mask`,
+    and is handy for multiple conditions and replacements.
+
+    Parameters
+    ----------
+    *args : array-like, scalar
+        Variable argument of tuples of conditions and expected replacements.
+        Takes the form:  `(condition0, replacement0)`,
+            `(condition1, replacement1)`, ... .
+        `condition` should be a 1-D boolean array.
+        When multiple boolean conditions are satisfied,
+        the first replacement is used.
+
+    default : scalar, array-like, default None
+        If provided, it is the replacement value to use
+        if all conditions evaluate to False.
+        If not specified, entries will be filled with the
+        corresponding NULL value (``np.nan`` for numpy dtypes,
+        ``pd.NA`` for extension dtypes).
+
+    level : int, default None
+        Alignment level if needed.
+
+        .. versionadded:: 2.2.0
+
+    Returns
+    -------
+    Series
+
+    See Also
+    --------
+    Series.mask : Replace values where the condition is True.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...     "a": [0,0,1,2],
+    ...     "b": [0,3,4,5],
+    ...     "c": [6,7,8,9]
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> pd.case_when((df.a.gt(0), df.a),   # condition, replacement
+    ...              (df.b.gt(0), df.b),   # condition, replacement
+    ...              default=df.c)        # optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+    from pandas._testing.asserters import assert_index_equal
+
+    # summary of logic -
+    # get booleans and replacements
+    # and iteratively get the final output via Series.mask
+    # iteration starts from the end,
+    # ensuring that if there are any booleans with the same replacement,
+    # the first one is taken
+    # as much as possible the computation is pushed to Series.mask
```

Comment:
I don't think this is necessary, and since it's somewhat divorced from the lines it's commenting on, it seems to me it has a risk of getting outdated.